### ISSUE TITLE
Feature Request: allow current/before `PRIMARY` to be specified in ERS

### ISSUE THREAD

**Issue by timvaillancourt** [vitessio/vitess#16430]
### Feature Description

This issue proposes that the current/before shard `PRIMARY` can be specified in ~~`PlannedReparentShard`~~ `EmergencyReparentShard` RPC requests in order to support a more idempotent(?) operation, where if the provided, "current" `PRIMARY` is no longer correct, the operation fails or no-ops

This could be implemented by adding an optional request-field `ExpectedPrimary *topodatapb.TabletAlias` containing the alias of who you think the `PRIMARY` is "now".  In the `EmergencyReparentShard` `.reparentShardLocked(...)` logic the current primary can be compared to this alias and a mismatch is handled

Your feedback is much appreciated!

### Use Case(s)

This support can prevent situations when an issue is fixed by VTOrc while external automation is also issuing `EmergencyReparentShard` operations

It would be great if the external automation could say "PRS only if alias X is still the `PRIMARY`" and if something changed the world in the process, this operation does no go through

**Comment by GuptaManan100** [comment]
VTOrc only runs `PRS` for a new uninitialized cluster. After this point, VTOrc never runs PRS. So I'm not sure if there are going to be that many collisions in external automation vs VTOrc running PRS.

**Comment by deepthi** [comment]
There are multiple ways of solving this without adding another flag to PRS
- external automation checks for the desired condition before issuing a PRS
- use AvoidPrimaryAlias as Manan suggested
Manan's point about such collisions being rare in practice is also valid. Perhaps you could elaborate on what situation you are running into and trying to solve.

**Comment by timvaillancourt** [comment]
@GuptaManan100 / @deepthi it looks like I made a mistake in relaying the problem internally. What we're looking for is adding the "expected" primary to `EmergencyReparentShard`, not `PlannedReparentShard` 🤦

To rephrase the scenario again, we would like the ability to specify the `PRIMARY` we expect to be the current primary in `EmergencyReparentShard` requests that importantly do not specify a candidate. This will allow external automation to be sure an ERS will happen only if the requestor's view of the world is still correct. If the `PRIMARY` changed in the time it took to issue an RPC, we'd like that request to fail instead of triggering a needless reparent

I've implemented this approach in this PR to illustrate: https://github.com/vitessio/vitess/pull/16852. Your thoughts are appreciated! 🙇
